diesel_derives/src/sql_function.rs

```diff
@@ -643,7 +728,61 @@ fn expand_nonvariadic(
             )
         };
 
-    quote! {
+    let (return_type_helper_module, return_type_helper_module_path) = if skip_return_type_helper {
+        (None, None)
+    } else {
+        let auto_derived_types = type_args
+            .iter()
+            .map(|type_arg| {
+                for arg in args {
+                    let Type::Path(path) = &arg.ty else {
+                        continue;
+                    };
+
+                    let Some(path_ident) = path.path.get_ident() else {
+                        continue;
+                    };
+
+                    if path_ident == type_arg {
+                        return Ok(arg.name.clone());
+                    }
+                }
+
+                Err(syn::Error::new(
+                    type_arg.span(),
+                    "cannot find argument corresponding to the generic",
+                ))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let arg_names_iter: Vec<_> = args.iter().map(|arg| arg.name.clone()).collect();
+
+        let return_type_module_name =
+            Ident::new(&format!("__{}_return_type", fn_name), fn_name.span());
+
+        let doc = format!("Return type of [`{fn_name}()`](super::{fn_name}()).");
```

Comment:
This link currently seems to link back on itself. We might want to try using a link disambiguator (e.g. `fn@super::{fn_name}()`) like described here: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html#namespaces-and-disambiguators

Also it might be worth to slightly tweak the wording here to 

```suggestion
        let doc = format!("Return type of the [`{fn_name}()`](fn@super::{fn_name}()) SQL function.");
```